File: mealplan/resources/rotation.py
"""The Sunday rotation: the dynamic list of themes Sunday cycles through.

Several rotations can exist (seasonal lists, experiments); ``activate``
stamps ``activated_at``, and new plans draw from the most recently activated
``active`` rotation — an action's effect belongs to this resource alone, so
activating one never mutates its siblings. New plans
auto-select that rotation and pre-theme their Sundays from it,
starting at ``data.position`` (the next Sunday's suggested theme —
``advance`` moves it after a theme gets used). A plan's ``set_sunday_theme``
guard reads this resource, so a Sunday theme not in the rotation is refused
with a remedy pointing at ``add_theme``.
"""
from __future__ import annotations

from enum import StrEnum

from pydantic import AwareDatetime, BaseModel, Field
from pydantic.json_schema import SkipJsonSchema

from waymark import Allow, Ctx, Deny, Resource, action, filterable, guard

DEFAULT_THEMES = ["breakfast for dinner", "indian", "greek", "soup night"]


class RotationState(StrEnum):
    ACTIVE = "active"      # the rotation plans draw from — at most one
    INACTIVE = "inactive"  # kept around (seasonal lists, experiments)


class RotationData(BaseModel):
    name: str = Field(default="Sunday rotation", min_length=1, max_length=100)
    themes: list[str] = Field(default_factory=lambda: list(DEFAULT_THEMES),
                              min_length=1)
    position: int = Field(default=0, ge=0,
                          description="Index of the next Sunday's theme")
    activated_at: AwareDatetime | None = Field(
        default=None, description="Set by `activate`; among active "
                                  "rotations, the most recent wins")


class RotationCreate(RotationData):
    """The create form: name it and list themes; the pointer starts at 0
    and only ever moves via ``advance``."""

    themes: list[str] = Field(default_factory=lambda: list(DEFAULT_THEMES),
                              min_length=1,
                              description="One theme per entry; Sunday cycles "
                                          "through these")
    position: SkipJsonSchema[int] = Field(default=0, ge=0)
    activated_at: SkipJsonSchema[AwareDatetime | None] = None


class ThemeInput(BaseModel):
    theme: str = Field(min_length=1, max_length=50)


@guard(else_="'{theme}' is the last theme left; the rotation cannot be empty.",
       vars=["theme"],
       # removable = what's on the rotation, unless that would empty it
       admits=("theme", lambda r: r.data.themes if len(r.data.themes) > 1 else []))
async def not_last_theme(r, inp: ThemeInput, ctx: Ctx) -> Allow | Deny:
    if len(r.data.themes) == 1 and inp.theme == r.data.themes[0]:
        return Deny(vars={"theme": inp.theme},
                    errors={"theme": ["cannot remove the last theme"]})
    return Allow()


class SundayRotation(Resource):
    kind = "rotation"
    State = RotationState
    Data = RotationData
    Create = RotationCreate

    initial = RotationState.INACTIVE

    summary = "{data.name} · {data.themes|len} themes · {state.label}"

    filterable = filterable(state=filterable.Eq | filterable.In)

    display = {"title": "{data.name}"}

    @action(from_=RotationState.INACTIVE, to=RotationState.ACTIVE,
            idempotent=True, reversible=False, confirm=False,
            display=dict(label="Make active", style="primary", order=1,
                         description="New plans draw Sunday themes from the "
                                     "most recently activated rotation"))
    async def activate(self, inp: None, ctx: Ctx) -> None:
        # an action's effect belongs to *this* resource, so activating never
        # touches siblings — among actives, the freshest activation wins
        self.data.activated_at = ctx.now

    @action(from_=RotationState.ACTIVE, to=RotationState.INACTIVE,
            idempotent=True, reversible=False, confirm=False,
            display=dict(label="Deactivate", order=4))
    async def deactivate(self, inp: None, ctx: Ctx) -> None:
        pass

    @action(from_=RotationState.ACTIVE, to=RotationState.ACTIVE,
            input=ThemeInput,
            idempotent=True, reversible=False, confirm=False,
            display=dict(label="Add theme", style="primary", order=1))
    async def add_theme(self, inp: ThemeInput, ctx: Ctx) -> None:
        if inp.theme not in self.data.themes:
            self.data.themes.append(inp.theme)

    @action(from_=RotationState.ACTIVE, to=RotationState.ACTIVE,
            input=ThemeInput, guards=[not_last_theme],
            idempotent=True, reversible=False, confirm=False,
            display=dict(label="Remove theme", order=2))
    async def remove_theme(self, inp: ThemeInput, ctx: Ctx) -> None:
        # removing an absent theme is a no-op, so retries stay replay-safe
        if inp.theme in self.data.themes:
            self.data.themes.remove(inp.theme)
            self.data.position %= len(self.data.themes)

    @action(from_=RotationState.ACTIVE, to=RotationState.ACTIVE,
            idempotent=False, reversible=False, confirm=False,
            display=dict(label="Next theme", order=3))
    async def advance(self, inp: None, ctx: Ctx) -> None:
        self.data.position = (self.data.position + 1) % len(self.data.themes)

File: mealplan10/src/mealplan10/resources/rotation.clj
(ns mealplan10.resources.rotation
  "The Sunday rotation: the dynamic list of themes Sunday cycles
  through.

  Several rotations can exist (seasonal lists, experiments); activate
  stamps activated_at, and new plans draw from the most recently
  activated active rotation — an action's effect belongs to this
  resource alone, so activating one never mutates its siblings. New
  plans auto-select that rotation and pre-theme their Sundays from it,
  starting at data.position (advance moves it after a theme gets
  used). A plan's set_sunday_theme guard reads this resource, so a
  Sunday theme not in the rotation is refused with a remedy pointing
  at add_theme.

  not_last_theme is a pure acceptance-set declaration — the set of
  removable themes is both the rendered enum and the enforcement;
  there is no separate check body to drift.

  Spelled in the batch-H declaration style: the whole machine is
  :flow rows (so :states is not spelled — the rows name them), and
  activate/deactivate declare each other as :undo — the engine
  verifies the pointers, so \"honestly reversible\" is graph-checked
  instead of a comment. Recorded deviations, each a sentence: advance
  spells its :safety explicitly (a flow row mints idempotent, and
  advance honestly is not — click again, move again); the theme edits
  keep :input (a bounded 50-char string has no field word). The law
  is unchanged; mealplan10.style-invariance-test pins the fingerprint
  hash byte-identical to the split spelling.

  Recorded deviation: v10 declares no field defaults, so the name and
  the starter themes land in :on-create when the create body leaves
  them blank."
  (:require [waymark10.dsl :refer [defresource defhandler guard]]))

(def default-themes
  ["breakfast for dinner" "indian" "greek" "soup night"])

;; removable = what's on the rotation, unless that would empty it.
;; One declaration: the rendered enum, the enforcement, and the
;; per-part availability all come from this set.
(def not-last-theme
  (guard {:name :not-last-theme
          :judges [:theme]
          :accepts (fn [row]
                     (let [themes (get-in row [:data :themes])]
                       (if (< 1 (count themes)) (vec themes) [])))
          :explain "'{theme}' cannot be removed; the rotation must keep at least one theme."}))

(defhandler activate-rotation [row _inp ctx]
  (assoc-in row [:data :activated_at] (:now ctx)))

(defhandler add-theme [row inp _ctx]
  (update-in row [:data :themes]
             (fn [themes]
               (if (some #{(:theme inp)} themes)
                 themes
                 (conj (vec themes) (:theme inp))))))

(defhandler remove-theme [row inp _ctx]
  ;; removing an absent theme is a no-op, so retries stay replay-safe
  (update row :data
          (fn [d]
            (if (some #{(:theme inp)} (:themes d))
              (let [themes (vec (remove #{(:theme inp)} (:themes d)))]
                (assoc d :themes themes
                       :position (mod (:position d) (count themes))))
              d))))

(defhandler advance-rotation [row _inp _ctx]
  (update row :data
          (fn [d]
            (assoc d :position (mod (inc (:position d))
                                    (count (:themes d)))))))

(defn- rotation-on-create
  "Blank create fields land their mealplan9 model defaults here."
  [row _ctx]
  (update row :data
          (fn [d]
            (-> d
                (update :name #(or % "Sunday rotation"))
                (update :themes #(if (seq %) (vec %) default-themes))
                (assoc :position 0)))))

(def theme-input
  [:map [:theme [:string {:min 1 :max 50}]]])

(defresource rotation
  {:kind :rotation
   :initial :inactive
   :summary "{data.name} · {state}"
   :schema [:map
            [:name [:string {:min 1 :max 100}]]
            [:themes [:vector {:min 1} [:string {:min 1 :max 50}]]]
            [:position [:int {:min 0}]]
            [:activated_at {:optional true} [:maybe :waymark/instant]]]
   ;; the create form: name it and list themes; the pointer starts at
   ;; 0 and only ever moves via advance
   :create-schema [:map
                   [:name {:optional true} [:maybe [:string {:min 1 :max 100}]]]
                   [:themes {:optional true}
                    [:maybe [:vector {:min 1} [:string {:min 1 :max 50}]]]]]
   :on-create rotation-on-create
   :filterable {:state #{:eq :in}}
   :display {:title "{data.name}"}
   ;; the whole machine as rows — the rows name the states, and the
   ;; activation pair name each other as the verified way back; the
   ;; self-loop list edits mint the idempotent-overwrite safety
   :flow
   [[:inactive :activate     :active
     {:undo :deactivate
      :handler activate-rotation
      :display {:label "Make active" :style :primary :order 1
                :description "New plans draw Sunday themes from the most recently activated rotation"}}]
    [:active   :deactivate   :inactive
     {:undo :activate
      :display {:label "Deactivate" :order 4}}]
    [:active   :add_theme    :active
     {:input theme-input
      :handler add-theme
      :display {:label "Add theme" :style :primary :order 1}}]
    [:active   :remove_theme :active
     {:input theme-input
      :requires [not-last-theme]
      :handler remove-theme
      :display {:label "Remove theme" :order 2}}]
    [:active   :advance      :active
     ;; honestly NOT idempotent — click again, move again — so the
     ;; row spells its safety instead of taking the minted overwrite
     {:safety {:idempotent false :reversible false :confirm false}
      :handler advance-rotation
      :display {:label "Next theme" :order 3}}]]})
